chore(flags): Use the new `ReadThroughCache` for the flags cache

haacked opened the pull request:
## Problem

We have a lot of code for reading/writing to the Redis cache and each code is different. The `ReadThroughCache` introduced in https://github.com/PostHog/posthog/pull/40751 encapsulates a good pattern for doing this.

## Changes

This uses the new `ReadThroughCache` to read and write the flags cache.

The TTL can be configured through the `FLAGS_CACHE_TTL_SECONDS` env var. The default remains `432000` or `5` days.

This also adds a new counter for an exceedingly rare situation (that we believe can no longer happen): 

```rust
counter!(
    FLAG_FILTER_DESERIALIZATION_ERROR_COUNTER, // "flags_filter_deserialization_error_total"
    "team_id" => row.team_id.to_string(),
    "flag_key" => row.key.clone(),
)
.increment(1);
```

## How did you test this code?

Unit Tests

Manually:

- [x] Test cache miss and cache hit
1. `redis-cli FLUSHALL`
2. `RUST_LOG=debug cargo run -p feature-flags` 
3. `curl -L 'http://localhost:8010/flags?v=2'...` (Verify response)
4. `redis-cli KEYS "posthog:1:team_feature_flags_*"` (confirm flags is cached for your team id)
5. `curl -L 'http://localhost:8010/flags?v=2'...` (Verify response, check logs for `Positive cache hit for key:`)

- [x] Test cache eviction at TTL
1. `redis-cli FLUSHALL`
2. `curl -L 'http://localhost:8010/flags?v=2'...` (Verify response)
3. `redis-cli TTL "posthog:1:team_feature_flags_{team_id}` (Verify key exists with TTL)
4. Wait for TTL
5. `curl -L 'http://localhost:8010/flags?v=2'...` (Verify response, check logs to for absence of `Positive cache hit for key:`)

- [x] Test response when Redis down
1. `redis-cli FLUSHALL`
2. `brew services stop redis`
3. `curl -L 'http://localhost:8010/flags?v=2'...` (Verify response)
4. `brew services restart redis`


👉 _Stay up-to-date with [PostHog coding conventions](https://posthog.com/docs/contribute/coding-conventions) for a smoother review._

## Changelog: (features only) Is this feature complete?

<!-- Optional, but helpful for our content team! -->
<!-- Yes if this is okay to go in the changelog. No if it's still hidden behind a feature flag, or part of a feature that's not complete yet, etc.  -->

haacked committed:
Use the new `ReadThroughCache` for the flags cache

This uses the new `ReadThroughCache` to read and write the flags cache.

haacked committed:
Move `FeatureFlagList::from_redis` to a test helper

This method is no longer needed. It's better left as a test helper.

haacked committed:
Track cache hits and misses

haacked commented on rust/feature-flags/src/flags/flag_service.rs:
```diff
@@ -141,56 +148,41 @@ impl FlagService {
+        let was_cache_hit = cache_result.was_cached();
+        if was_cache_hit {
+            inc(FLAG_CACHE_HIT_COUNTER, &[], 1);
+        }
```
Good call. I pushed a fix.

dmarticus reviewed (approved):
two comments, non-blocking, generally looks great!

haacked committed:
Add clarifying comment

haacked committed:
Return `CacheResult<FlagsWithMetadata>` more directly

dmarticus commented on rust/feature-flags/src/flags/test_helpers.rs:
```diff
@@ -62,3 +68,40 @@ pub fn create_simple_flag(properties: Vec<PropertyFilter>, rollout_percentage: f
+///
+/// This bypasses the ReadThroughCache and directly reads from Redis,
+/// useful for testing cache behavior and verifying cache contents.
+pub async fn get_flags_from_redis(
```
bother, should've checked!

haacked committed:
Move update_flags_in_redis to test helpers

This function is only used in tests to set up test data by directly
writing flags to Redis, bypassing the ReadThroughCache. Moving it to
test_helpers.rs makes its purpose clear and keeps test-only code
separate from production code.

haacked committed:
Add back-compat test

haacked committed:
Ensure we can read existing cache entries

haacked committed:
Refactor: Simplify method

haacked committed:
Remove TODO comment

haacked committed:
Be more defensive
